README.md
# Linux Hub for Workroom project

Soon to be transferred here from our private repositories...


CHANGELOG.adoc
= Change Log
:gh-url: https://github.com/bigclownlabs/bc-workroom-hub

== link:{gh-url}/tree/v1.2.2[v1.2.2] (2017-01-24)
Fixed::
* Fix gateway to not log every received line from serial port on INFO level.

== link:{gh-url}/tree/v1.2.1[v1.2.1] (2017-01-23)
Fixed::
* Fix bc-common debian package to create user and group `bigclown`

== link:{gh-url}/tree/v1.2.0[v1.2.0] (2017-01-21)
Added::
* Add option `--list` to bc-workroom-gateway
* Led-strip plugin confirm set data
* Add publish errors in led-strip plugin
* Add get data to led-strip plugin
* Add recalculate config rules if change led-strip mode

Changed::
* Blynk plugin change brightness range to 0..1023
* Blynk plugin change color range to [0..1023, 0..1023, 0..1023]

Fixed::
* Fix condition for the default rule in led-strip plugin

== link:{gh-url}/tree/v1.1.0[v1.1.0] (2017-01-12)
Changed::
* Change device symlink created by udev from `/dev/bcbridgehid%n` to `/dev/ft260hidraw%n`.

Fixed::
* Fix postinst scripts to not wait for a service to be started.

== link:{gh-url}/tree/v1.0.3[v1.0.3] (2016-01-11)
Added::
* Udev rules for bc-bridge

== link:{gh-url}/tree/v1.0.2[v1.0.2] (2016-12-30)
Fixed::
* Fix check_config in plugin lef-strip

== link:{gh-url}/tree/v1.0.1[v1.0.1] (2016-12-19)
Fixed::
* Reapire build script

== link:{gh-url}/tree/v1.0.0[1.0.0] (2016-12-19)

* The first stable release.


blynk/main.js
#!/usr/bin/env nodejs
"use strict";
/**
 * Plugin: for http://www.blynk.cc/
 *
 * virtual pin 0 - temperature
 * virtual pin 1 - relative-humidity
 * virtual pin 2 - set brightness range 0 .. 255
 * virtual pin 3 - light
 * virtual pin 4 - relay
 * virtual pin 5 - zeRGBa widget
 *
 * Example config:
 * mosquitto_pub -t "plugin/blynk/config" -m '{"token":"your token id"}' -r
 */

var doc = "BigClown plugin for http://www.blynk.cc/.\n\
\n\
Usage:\n\
  bc-workroom-blynk [options]\n\
  bc-workroom-blynk --help\n\
\n\
Options:\n\
  -D --debug                   Print debug messages.\n\
  -h HOST --host=HOST          MQTT host to connect to (default is localhost).\n\
  -p PORT --port=PORT          MQTT port to connect to (default is 1883).\n\
  -t TOPIC --base-topic=TOPIC  Base MQTT topic (default is nodes).\n\
  -v --version                 Print version.\n\
  --help                       Show this message.\n\
";
var VERSION = require('./package.json').version;

var mqtt = require('mqtt');
var blynkLib = require('blynk-library');
var docopt = require('docopt');
var Log = require('log');

var options = docopt.docopt(doc, {version: 'bc-workroom-blynk ' + VERSION});

var logging = new Log(options['--debug'] ? 'debug' : 'info');

var DEFAULT_MQTT_HOST = 'localhost';
var DEFAULT_MQTT_PORT = 1883;
var DEFAULT_MQTT_TOPIC = 'blynk';

var brokerUrl = "mqtt://" + (options['--host'] || DEFAULT_MQTT_HOST) + ":" + (options['--port'] || DEFAULT_MQTT_PORT)
var client = mqtt.connect(brokerUrl);
var blynkClient = null;
var pluginTopic = "plugin/" + (options['--base-topic'] || DEFAULT_MQTT_TOPIC);


function BlynkClient(token, client) {
  this.token = token;

  var blynk = new blynkLib.Blynk(token);

  var v0 = new blynk.VirtualPin(0);
  var v1 = new blynk.VirtualPin(1);
  var v2 = new blynk.VirtualPin(2);
  var v3 = new blynk.VirtualPin(3);
  var v4 = new blynk.VirtualPin(4);
  var v5 = new blynk.VirtualPin(5);

  var topics = ["nodes/remote/+/+", "nodes/base/light/-", "plugin/led-strip/data/set/ok", "plugin/led-strip/data"];

  function addSubscibe() {
    for(var topic of topics) {
        logging.debug("mqtt subscribe", topic);
        client.subscribe(topic);
    }

    client.publish("nodes/base/light/-/get", "{}");
    client.publish("nodes/base/relay/-/get", "{}");
    client.publish("plugin/led-strip/data/get", "{}");
  }

  v2.on('write', function(value) {
    logging.debug('blynk v2', 'on write', value[0]);
    var brightness = Math.floor(parseInt(value[0]) / 1023 * 255);
    client.publish("plugin/led-strip/data/set", JSON.stringify({ "brightness": brightness}), {retain: true} );
  });

  v3.on('write', function(value) {
    logging.debug('blynk v3', 'on write', value[0]);
    client.publish("nodes/base/light/-/set", JSON.stringify({ "state": value[0] == "1" ? true : false }));
  });

  v4.on('write', function(value) {
    logging.debug('blynk v4', 'on write', value[0]);
    client.publish("nodes/base/relay/-/set", JSON.stringify({ "state": value[0] == "1" ? true : false }));
  });

  v5.on('write', function(value) {
    logging.debug('blynk v5', 'on write', value);

    var payload = {"state": "rules"};

    if ((value[0] !== '1023') || (value[1] !== '1023') && (value[2] !== '1023')) {
      var red = Math.floor((parseInt(value[0]) || 0) / 1023 * 255);
      var green = Math.floor((parseInt(value[1]) || 0) / 1023 * 255);
      var blue = Math.floor((parseInt(value[2]) || 0) / 1023 * 255);
      payload = {"color": [red, green, blue, 0], "state": "color"};
    }

    client.publish("plugin/led-strip/data/set", JSON.stringify(payload));
  });

  function clientMessage(topic, message) {
    logging.debug("BlynkClient message", topic, message.toString());

    if (!message) return;

    try {
      var payload = JSON.parse(message.toString());

      if ((topic === "nodes/remote/thermometer/i2c0-49") || topic === "nodes/remote/thermometer/i2c1-49") {
        if (payload['temperature']) {
          v0.write(payload['temperature'][0]);
        }

      } else if ((topic === "nodes/remote/humidity-sensor/i2c0-40") || (topic === "nodes/remote/humidity-sensor/i2c1-40") ) {
        if (payload['relative-humidity']) {
          v1.write(payload['relative-humidity'][0]);
        }

      } else if (topic === "nodes/base/light/-") {
        v3.write(payload['state'] ? "1" : "0");

      } else if (topic === "nodes/base/relay/-") {
        v4.write(payload['state'] ? "1" : "0");

      } else if ((topic === "plugin/led-strip/data/set/ok") || (topic === "plugin/led-strip/data")) {
        if ('brightness' in payload) {
          v2.write(Math.floor(payload['brightness'] / 255 * 1023))
        }
        if ('color' in payload) {
          var red = Math.floor(payload['color'][0] / 255 * 1023);
          var green = Math.floor(payload['color'][1] / 255 * 1023);
          var blue = Math.floor(payload['color'][2] / 255 * 1023);
          v5.write([red, green, blue]);
        }
      }

    } catch (error) {
      logging.error('Error', error, topic, message.toString());
      return;
    }
  }

  client.on('message', clientMessage);
  client.on('connect', addSubscibe);
  blynk.on('connect', addSubscibe);
  blynk.on('error', function(e){
    logging.error('blynk', e);
    client.publish(pluginTopic + "/config/error", JSON.stringify({msg:e}));
  });

  this.destroy = function() {
    blynk.removeListener('connect', addSubscibe);
    clearInterval(blynk.timerConn);
    client.removeListener('connect', addSubscibe);
    client.removeListener('message', clientMessage);
    for(var topic of topics) {
        client.unsubscribe(topic);
    }
    blynk.disconnect(false);
  }
}

client.on('connect', function() {
  logging.info("mqtt connect");
  logging.debug("mqtt subscribe", pluginTopic + "/config");
  client.subscribe(pluginTopic + "/config")
});

client.on('message', function(topic, message) {

  if (topic === pluginTopic + "/config") {
    var payload = JSON.parse(message.toString());
    logging.debug(topic, payload);

    if (payload.token) {
      if (!blynkClient) {
        logging.info("start BlynkClient");
        blynkClient = new BlynkClient(payload.token, client);

      } else if (blynkClient.token !== payload.token) {
        logging.info("destroy BlynkClient");
        blynkClient.destroy();
        logging.info("start BlynkClient");
        blynkClient = new BlynkClient(payload.token, client);

      }
    }
  }

});

client.on('disconnect', function() {
  logging.info("mqtt disconnect");
})


script/utils.sh
# vim: set ts=4:

: ${BASE_DIR:="$(pwd)"}
: ${BUILD_DIR:="$(pwd)/build"}
: ${TARGET_DIR:="$(pwd)/dist"}


die() {
	# bold red
	printf '\033[1;31mERROR:\033[0m %s\n' "$1" >&2
	shift
	printf '  %s\n' "$@"
	exit 2
}

einfo() {
	# bold cyan
	printf '\033[1;36m> %s\033[0m\n' "$@" >&2
}

ewarn() {
	# bold yellow
	printf '\033[1;33m> %s\033[0m\n' "$@" >&2
}

# Infers project version in Debian format from v* tags in git repository.
# Underscores (_) and hyphens (-) are replaced with tilde (~) - the suffix
# separator in Debian version number format.
# If HEAD is not tagged (i.e. this is not a release), then it returns last
# version with suffix ~dev.
pkgver_from_git() {
	local desc

	# If git HEAD is tagged as v*, then it prints name of the tag.
	if desc="$(git describe --tags --exact-match --match 'v*' 2>/dev/null)"; then
		echo "${desc#v}" | tr '[_-]' '~'
	# Prints name of the last v* tag in the tree with suffix `-<n>-<abbrev>`.
	elif desc="$(git describe --tags --match 'v*' 2>/dev/null)"; then
		# Replace git suffix with ~dev.
		echo "$desc" | sed -En 's/^v([^-]+).*/\1~dev/p'
	else
		return 1
	fi
}

has() {
	command -v "$1" >/dev/null 2>&1
}

sed_inplace() {
	if sed --version >/dev/null 2>&1; then
		sed -r -i'' "$@"  # GNU or Busybox sed
	else
		sed -E -i '' "$@"  # BSD sed
	fi
}


gateway/bc-workroom-gateway.py
#!/usr/bin/env python3

"""
BigClown gateway between Base unit connected via USB and MQTT broker.

Usage:
  bc-workroom-gateway [options]

Options:
  -D --debug                   Print debug messages.
  -d DEVICE --device=DEVICE    Path to Base unit's device node (default is /dev/ttyACM0).
  -h HOST --host=HOST          MQTT host to connect to (default is localhost).
  -p PORT --port=PORT          MQTT port to connect to (default is 1883).
  -t TOPIC --base-topic=TOPIC  Base MQTT topic (default is nodes).
  --list                       Show list of available devices
  -v --version                 Print version.
  --help                       Show this message.
"""

import os
import sys
from logging import DEBUG, INFO
import logging as log
import json
import platform

from docopt import docopt
import paho.mqtt.client as mqtt
from serial import Serial
from serial.tools import list_ports

if platform.system() == 'Linux':
    import fcntl

__version__ = '@@VERSION@@'

DEFAULT_MQTT_HOST = 'localhost'
DEFAULT_MQTT_PORT = 1883
DEFAULT_MQTT_TOPIC = 'nodes'
DEFAULT_DEVICE = '/dev/ttyACM0'

LOG_FORMAT = '%(asctime)s %(levelname)s: %(message)s'


def mgtt_on_connect(client, userdata, flags, rc):
    log.info('Connected to MQTT broker with (code %s)', rc)

    for subtopic in ('base/+/+/+', 'base/+/+/+/+'):
        client.subscribe(userdata['base_topic'] + subtopic)


def mgtt_on_message(client, userdata, msg):
    subtopic = msg.topic[len(userdata['base_topic']):]
    userdata['serial'].write(b'["' + subtopic.encode('utf-8') + b'",' + msg.payload + b']\n')


def main():
    arguments = docopt(__doc__, version='bc-workroom-gateway %s' % __version__)
    opts = {k.lstrip('-').replace('-', '_'): v
            for k, v in arguments.items() if v}

    log.basicConfig(level=DEBUG if opts.get('debug') else INFO, format=LOG_FORMAT)

    if opts.get('list'):
        for p in list_ports.comports():
            print(p)
        return

    serial = Serial(opts.get('device', DEFAULT_DEVICE), timeout=3.0)
    if platform.system() == 'Linux':
        fcntl.flock(serial.fileno(), fcntl.LOCK_EX | fcntl.LOCK_NB)
        log.debug("flock %d" % serial.fileno())

    serial.write(b'\n')

    base_topic = opts.get('base_topic', DEFAULT_MQTT_TOPIC).rstrip('/') + '/'

    mqttc = mqtt.Client(userdata={'serial': serial, 'base_topic': base_topic})
    mqttc.on_connect = mgtt_on_connect
    mqttc.on_message = mgtt_on_message

    mqttc.connect(opts.get('host', DEFAULT_MQTT_HOST),
                  opts.get('port', DEFAULT_MQTT_PORT),
                  keepalive=10)
    mqttc.loop_start()

    while True:
        line = serial.readline()
        if line:
            log.debug(line)
            try:
                talk = json.loads(line.decode())
                mqttc.publish(base_topic + talk[0], json.dumps(talk[1]), qos=1)
            except:
                log.error('Received malformed message: %s', line)


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(0)
    except Exception as e:
        log.error(e)
        if os.getenv('DEBUG', False):
            raise e
        sys.exit(1)


ledstrip/bc-workroom-ledstrip.py
#!/usr/bin/env python3

"""
BigClown plugin for controlling color of an LED strip based on temperature
and relative humidity read from MQTT.

MQTT topic: plugin/led-strip/config

Usage:
  bc-workroom-ledstrip [options]
  bc-workroom-ledstrip --help

Options:
  -D --debug           Print debug messages.
  -h HOST --host=HOST  MQTT host to connect to (default is localhost).
  -p PORT --port=PORT  MQTT port to connect to (default is 1883).
  -v --version         Print version.
  --help               Show this message.
"""

import base64
import json
import logging as log
from logging import DEBUG, INFO
import os
import sys
import time

from docopt import docopt
import paho.mqtt.client as mqtt

__version__ = '@@VERSION@@'

DEFAULT_MQTT_HOST = 'localhost'
DEFAULT_MQTT_PORT = 1883

DEFAULT_PLUGIN_CONFIG = {
    'values': [
        {'relative-humidity': {'from': 60}, 'color': [0, 255, 0, 0]},
        {'relative-humidity': {'to': 30}, 'color': [255, 255, 0, 0]},
        {'temperature': {'from': 26}, 'color': [255, 0, 0, 0]},
        {'temperature': {'to': 22}, 'color': [0, 0, 255, 0]},
        {'color': [0, 0, 0, 255]},
    ],
    'brightness': 255,
}

LOG_FORMAT = '%(asctime)s %(levelname)s: %(message)s'


# TODO: refactor!
def check_config(config):
    check_config_key = {'values', 'brightness'}
    config_key = set(config.keys())
    if config_key - check_config_key:
        raise AttributeError('unknown key(s): ' + ', '.join(config_key - check_config_key))
    if check_config_key - config_key:
        raise AttributeError('missing key(s): ' + ', '.join(check_config_key - config_key))

    if config['brightness'] < 0 or config['brightness'] > 255:
        raise AttributeError('brightness must be in range 0..255')

    if not isinstance(config['values'], list):
        raise AttributeError('values must be of type list')

    check_row_keys = {'color', 'temperature', 'relative-humidity', 'label'}

    for row in config['values']:
        row_keys = set(row.keys())

        if len(row_keys) == 0:
            raise AttributeError('%s is empty' % row)

        if row_keys - check_row_keys:
            raise AttributeError('unknown key: ' + ', '.join(row_keys - check_row_keys))

        if 'color' not in row:
            raise AttributeError('missing key: color')

        if len(row['color']) != 4 or len(list(filter(lambda x: x >= 0 <= 255, row['color']))) != 4:
            raise AttributeError('bad define color %s' % row['color'])

        for key in ('relative-humidity', 'temperature'):
            if key in row:
                if not isinstance(row[key], dict):
                    raise AttributeError('key %s is not type object' % key)

                for k, v in row[key].items():
                    if k not in ('from', 'to'):
                        raise AttributeError('unknown key: %s' % k)
                    elif not isinstance(v, (int, float)):
                        raise AttributeError(
                            'value %s of key %s is not type int or float' % (repr(v), k))


def make_pixels(red, green, blue, white, brightness=255, nums=144):
    pixel = [(red * (brightness + 1)) >> 8,
             (green * (brightness + 1)) >> 8,
             (blue * (brightness + 1)) >> 8,
             (white * (brightness + 1)) >> 8]
    log.debug('Pixel: %s', pixel)

    buf = pixel * nums
    pixels = base64.b64encode(bytearray(buf)).decode()
    assert(len(pixels) == 768)

    return pixels


# TODO: refactor!
def set_led_strip(client, userdata):
    now = time.time()

    if (now - userdata.get('temperature-ts', 0)) > 180:
        userdata['temperature'] = None

    if (now - userdata.get('relative-humidity-ts', 0)) > 180:
        userdata['relative-humidity'] = None

    color = None

    for row in userdata['config']['values']:
        valid = True

        for key in ('relative-humidity', 'temperature'):
            if valid and key in row:

                measured_value = userdata.get(key, None)
                if measured_value is None:
                    valid = False
                    break

                for key, value in row[key].items():
                    if key == 'to' and measured_value <= value:
                        pass
                    elif key == 'from' and measured_value >= value:
                        pass
                    else:
                        valid = False

        if valid:
            log.debug(row)
            color = row['color']
            break

    if color:
        pixels = make_pixels(*color, brightness=userdata['config']['brightness'])
        client.publish('nodes/base/led-strip/-/set', json.dumps({'pixels': pixels}))


def mgtt_on_connect(client, userdata, flags, rc):
    log.info('Connected to MQTT broker (code %s)', rc)

    for topic in ('plugin/led-strip/config',
                  'plugin/led-strip/config/set',
                  'nodes/remote/+/+',
                  'nodes/base/push-button/-'):
        client.subscribe(topic)


def mgtt_on_message(client, userdata, msg):
    payload = msg.payload.decode('utf-8')

    log.debug('%s: %s', msg.topic, payload)

    if msg.topic == 'plugin/led-strip/config' and not payload:
        # on empty payload load deafault config
        payload = DEFAULT_PLUGIN_CONFIG
    else:
        try:
            payload = json.loads(payload)
        except ValueError as e:
            log.error('Malformed payload: %s', e)
            return

    now = time.time()

    if msg.topic == 'nodes/remote/thermometer/i2c0-49':
        try:
            userdata['temperature'] = float(payload['temperature'][0])
            userdata['temperature-ts'] = now
        except (TypeError, ValueError) as e:
            log.error('Invalid temperature: %s', e)

    elif msg.topic == 'nodes/remote/humidity-sensor/i2c0-40':
        try:
            userdata['relative-humidity'] = float(payload['relative-humidity'][0])
            userdata['relative-humidity-ts'] = now
        except (TypeError, ValueError) as e:
            log.error('Invalid relative-humidity: %s', e)

    elif msg.topic == 'plugin/led-strip/config':
        try:
            check_config(payload)
            userdata['config'] = payload
        except Exception as e:
            log.error('Invalid config: %s', e)

    elif msg.topic == 'plugin/led-strip/config/set':
        if 'brightness' in payload:
            config = userdata['config']
            config['brightness'] = payload['brightness']
            try:
                check_config(config)
                client.publish('plugin/led-strip/config', json.dumps(config), retain=True)
            except Exception as e:
                log.error('Invalid config: %s', e)
            return

    set_led_strip(client, userdata)


def main():
    arguments = docopt(__doc__, version='bc-workroom-ledstrip %s' % __version__)
    opts = {k.lstrip('-').replace('-', '_'): v
            for k, v in arguments.items() if v}

    log.basicConfig(level=DEBUG if opts.get('debug') else INFO, format=LOG_FORMAT)

    check_config(DEFAULT_PLUGIN_CONFIG)

    client = mqtt.Client(userdata={'config': DEFAULT_PLUGIN_CONFIG})
    client.on_connect = mgtt_on_connect
    client.on_message = mgtt_on_message

    client.connect(opts.get('host', DEFAULT_MQTT_HOST),
                   opts.get('port', DEFAULT_MQTT_PORT),
                   keepalive=10)
    client.loop_forever()


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(0)
    except Exception as e:
        log.error(e)
        if os.getenv('DEBUG', False):
            raise e
        sys.exit(1)


led-strip/bc-workroom-led-strip.py
#!/usr/bin/env python3

"""
BigClown plugin for controlling color of an LED strip based on temperature
and relative humidity read from MQTT.

MQTT topic: plugin/led-strip/config

Usage:
  bc-workroom-led-strip [options]
  bc-workroom-led-strip --help

Options:
  -D --debug           Print debug messages.
  -h HOST --host=HOST  MQTT host to connect to (default is localhost).
  -p PORT --port=PORT  MQTT port to connect to (default is 1883).
  -v --version         Print version.
  --help               Show this message.
"""

import base64
import json
import logging as log
from logging import DEBUG, INFO
import os
import sys
import time

from docopt import docopt
import paho.mqtt.client as mqtt

__version__ = '@@VERSION@@'

DEFAULT_MQTT_HOST = 'localhost'
DEFAULT_MQTT_PORT = 1883

DEFAULT_PLUGIN_CONFIG = {
    'rules': [
        {'relative-humidity': {'from': 60}, 'color': [0, 255, 0, 0]},
        {'relative-humidity': {'to': 30}, 'color': [255, 255, 0, 0]},
        {'temperature': {'from': 26}, 'color': [255, 0, 0, 0]},
        {'temperature': {'to': 22}, 'color': [0, 0, 255, 0]},
        {'color': [0, 0, 0, 230]}
    ]
}

DEFAULT_PLUGIN_DATA = {
    'brightness': 255,
    "state": "rules",
    'color': [255, 0, 0, 0]
}

LOG_FORMAT = '%(asctime)s %(levelname)s: %(message)s'


def check_color_format(color):
    if len(color) != 4:
        raise AttributeError('bad define color %s' % color)
    for c in color:
        if not isinstance(c, int) or c < 0 or c > 255:
            raise AttributeError('bad define color %s' % color)
    return color


def check_config(config):
    # compatibility with old implementation
    if 'values' in config:
        config['rules'] = config['values']
        del config['values']
    if 'brightness' in config:
        del config['brightness']

    if tuple(config) != ('rules',):
        raise AttributeError('missing or unknown key')

    if not isinstance(config['rules'], list):
        raise AttributeError('values must be of type list')

    check_row_keys = {'color', 'temperature', 'relative-humidity', 'label'}

    for row in config['rules']:
        if not row:
            raise AttributeError('not allowed empty rule')

        row_keys = set(row.keys())

        if row_keys - check_row_keys:
            raise AttributeError('in rule unknown key: ' + ', '.join(row_keys - check_row_keys))

        if 'color' not in row:
            raise AttributeError('in rule missing key: color')

        check_color_format(row['color'])

        for key in ('relative-humidity', 'temperature'):
            if key in row:
                if not isinstance(row[key], dict):
                    raise AttributeError('key %s is not type object' % key)

                for k, v in row[key].items():
                    if k not in ('from', 'to'):
                        raise AttributeError('in rule unknown key: %s' % k)
                    elif not isinstance(v, (int, float)):
                        raise AttributeError(
                            'in rule: value %s of key %s is not type int or float' % (repr(v), k))
    return config


def make_pixels(red, green, blue, white=None, brightness=255, count=144):
    pixel = [(red * (brightness + 1)) >> 8,
             (green * (brightness + 1)) >> 8,
             (blue * (brightness + 1)) >> 8]
    if white is not None:
        pixel.append((white * (brightness + 1)) >> 8)

    log.debug('Pixel: %s', pixel)

    buf = pixel * count
    pixels = base64.b64encode(bytearray(buf)).decode()

    log.debug('len(base64): %d', len(pixels))

    return pixels


def base_led_strip_set_pixels(client, userdata):
    color = None

    if userdata['data']['state'] == 'rules':
        now = time.time()

        if (now - userdata.get('temperature-ts', 0)) > 180:
            userdata['temperature'] = None

        if (now - userdata.get('relative-humidity-ts', 0)) > 180:
            userdata['relative-humidity'] = None

        for row in userdata['config']['rules']:

            for key in ('relative-humidity', 'temperature'):
                rule = row.get(key, None)
                if rule is None:
                    continue

                measured_value = userdata.get(key, None)
                if measured_value is None:
                    break

                if rule:
                    if 'to' in rule and measured_value > rule['to']:
                        break

                    if 'from' in rule and measured_value < rule['from']:
                        break

            else:
                log.debug(row)
                color = row['color']
                break

    elif userdata['data']['state'] == 'color':
        color = userdata['data']['color']

    if color:
        if userdata['mode'] == 'rgb':
            color = color[:3]
        pixels = make_pixels(*color,
                             brightness=userdata['data']['brightness'],
                             count=userdata['count'])
        client.publish('nodes/base/led-strip/-/set', json.dumps({'pixels': pixels}))


def mgtt_on_connect(client, userdata, flags, rc):
    log.info('Connected to MQTT broker (code %s)', rc)

    for topic in ('plugin/led-strip/config',
                  'plugin/led-strip/data/+',
                  'nodes/remote/+/+',
                  'nodes/base/+/+',
                  'nodes/base/led-strip/-/config/set',
                  # compatibility with old implementation
                  'plugin/led-strip/config/set',
                  'plugin/led-strip/set'):
        log.debug('subscribe %s', topic)
        client.subscribe(topic)


def mgtt_on_message(client, userdata, msg):
    payload = msg.payload.decode('utf-8')

    log.debug('%s: %s', msg.topic, payload)

    if msg.topic == 'plugin/led-strip/config' and not payload:
        # on empty payload load deafault config
        payload = DEFAULT_PLUGIN_CONFIG
    else:
        try:
            payload = json.loads(payload)
        except ValueError as e:
            log.error('Malformed payload: %s', e)
            return

    now = time.time()

    # compatibility with old implementation
    if msg.topic == 'plugin/led-strip/set' and 'color' in payload:
        msg.topic = 'plugin/led-strip/data/set'
    elif msg.topic == 'plugin/led-strip/config/set' and 'brightness' in payload:
        msg.topic = 'plugin/led-strip/data/set'

    if msg.topic in ('nodes/remote/thermometer/i2c0-49',
                     'nodes/remote/thermometer/i2c1-49'):
        try:
            userdata['temperature'] = float(payload['temperature'][0])
            userdata['temperature-ts'] = now
        except (TypeError, ValueError) as e:
            log.error('Invalid temperature: %s', e)
            client.publish(msg.topic + '/-/error',
                           json.dumps({'msg': 'Invalid temperature: %s' % e}))

    elif msg.topic in ('nodes/remote/humidity-sensor/i2c0-40',
                       'nodes/remote/humidity-sensor/i2c1-40'):
        try:
            userdata['relative-humidity'] = float(payload['relative-humidity'][0])
            userdata['relative-humidity-ts'] = now
        except (TypeError, ValueError) as e:
            log.error('Invalid relative-humidity: %s', e)
            client.publish(msg.topic + '/-/error',
                           json.dumps({'msg': 'Invalid relative-humidity: %s' % e}))

    elif msg.topic == 'plugin/led-strip/data/set':
        try:
            ok_payload = {}

            if 'color' in payload:
                ok_payload['color'] = check_color_format(payload['color'])

            if 'brightness' in payload:
                brightness = int(payload['brightness'])
                if brightness < 0 or brightness > 255:
                    raise Exception('brightness out of range')

                ok_payload['brightness'] = brightness

            if 'state' in payload:
                if payload['state'] not in ('rules', 'color', 'disable'):
                    raise ValueError('state values is rules or color')
                ok_payload['state'] = payload['state']

            if ok_payload:
                userdata['data'].update(ok_payload)
                client.publish('plugin/led-strip/data/set/ok', json.dumps(ok_payload))

        except Exception as e:
            log.error('Invalid data: %s', e)
            client.publish(msg.topic + '/error', json.dumps({"msg": 'Invalid data: ' + str(e)}))
            return

    elif msg.topic == 'plugin/led-strip/data/get':
        client.publish('plugin/led-strip/data', json.dumps(userdata['data']))

    elif msg.topic == 'nodes/base/led-strip/-/config/set':
        try:
            if payload.get('mode', None) not in ('rgb', 'rgbw'):
                raise ValueError('mode values is rgb or rgbw')

            userdata['mode'] = payload['mode']
            userdata['count'] = int(payload['count'])

            rules = userdata['config']['rules']
            for rule in rules:
                if userdata['mode'] == 'rgb':
                    if rule['color'][:3] == [0, 0, 0]:
                        rule['color'] = [rule['color'][3]] * 4
                else:
                    if rule['color'] == [rule['color'][3]] * 4:
                        rule['color'] = [0, 0, 0, rule['color'][3]]

            client.publish('plugin/led-strip/config', json.dumps(userdata['config']), retain=True)

        except (TypeError, ValueError) as e:
            log.error('Invalid led-strip config: %s', e)
            client.publish(msg.topic + '/error',
                           json.dumps({'msg': 'Invalid led-strip config: %s' % e}))
            return

    elif msg.topic == 'plugin/led-strip/config':
        try:
            userdata['config'] = check_config(payload)
        except Exception as e:
            log.error('Invalid config: %s', e)
            client.publish(msg.topic + '/-/error', json.dumps({"msg": 'Invalid config: ' + str(e)}))
            return

    base_led_strip_set_pixels(client, userdata)


def main():
    arguments = docopt(__doc__, version='bc-workroom-led-strip %s' % __version__)
    opts = {k.lstrip('-').replace('-', '_'): v
            for k, v in arguments.items() if v}

    log.basicConfig(level=DEBUG if opts.get('debug') else INFO, format=LOG_FORMAT)

    client = mqtt.Client(userdata={
        'config': check_config(DEFAULT_PLUGIN_CONFIG),
        'data': DEFAULT_PLUGIN_DATA,
        'count': 144,
        'mode': 'rgbw'})
    client.on_connect = mgtt_on_connect
    client.on_message = mgtt_on_message

    client.connect(opts.get('host', DEFAULT_MQTT_HOST),
                   opts.get('port', DEFAULT_MQTT_PORT),
                   keepalive=10)
    client.loop_forever()


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        sys.exit(0)
    except Exception as e:
        log.error(e)
        if os.getenv('DEBUG', False):
            raise e
        sys.exit(1)
